### ISSUE TITLE
Add coerce_types flag to parquet ArrowWriter

### ISSUE THREAD

**Issue by tustvold** [apache/arrow-rs#1938]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

As discussed in #1666 not all types can be represented within a parquet schema.

**Describe the solution you'd like**

The consensus appears to be to:

* By default faithfully round-trip the source data, performing no potentially lossy type conversion
* Add a coerce_types flag that will use the arrow cast kernels to coerce incompatible types prior to writing them

In particular

### Date64

If not coerce_types, write as Int64 and embed logical type in arrow schema only. Otherwise case to Date32

### Timestamp

If not coerce_types, write as is, setting LogicalType / ConvertedType only where appropriate.

If coerce_types, cast to a UTC timestamp with the closest supported time unit, likely needing #1936.

### Interval

If not coerce_types, write as FixedSizeBinaryArray matching the arrow representation and store logical type in arrow schema.

If coerce_types, convert to the relevant parquet representation.

**Describe alternatives you've considered**

See #1666

**Comment by dsgibbons** [comment]
take

**Comment by alamb** [comment]
I think this issue was done in https://github.com/apache/arrow-rs/pull/6840

If that is not correct, please reopen / let me know what else needs to be done

**Comment by dsgibbons** [comment]
@alamb my original PR only handled Date64. I'm not sure if Interval and Timestamp are still outstanding.

**Comment by alamb** [comment]
reopening per @dsgibbons 's comments

**Comment by alamb** [comment]
@dsgibbons  would you be willing to make a PR to complete Interval and Timestamp so we can close this issue?

**Comment by dsgibbons** [comment]
@alamb yes, but will probably take a while (1-2 months). If someone else wants to finish this off in the meantime then go ahead. I'll comment once start working on this.

**Comment by alamb** [comment]
Thank you @dsgibbons !

**Comment by nssalian** [comment]
@alamb, is this an active issue? I'm happy to take this on if no one else is working on it at the moment.

**Comment by alamb** [comment]
I am not sure -- you will need to research if there is any remaining items to do

Thank you @nssalian 🙏

**Comment by dsgibbons** [comment]
I have not done any work on this @nssalian, so feel free to take this.

**Comment by CuteChuanChuan** [comment]
Hi @alamb , could I give it a try?

**Comment by alamb** [comment]
Please do -- thanks @CuteChuanChuan
